### PR TITLE
Adam/replay from gen

### PR DESCRIPTION
**Pull request by czarcas7ic**
* Added rocksdb
* Added archive node as persistent peer to kickstart first blocks
* Inc to 64GB requirement with swap
* Git stash to prevent losing peoples work if they made local changes

Need to eventually go through and figure out which logs to squelch and which ones not to

### CODE REVIEW

File: osmosis-installer.py [hunk 1]
```diff
@@ -582,10 +750,12 @@ def initSetup ():
         os.chdir(os.path.expanduser(HOME+'/osmosis'))
         if networkAns == "1":
             print(bcolors.OKGREEN + "(5/5) Installing Osmosis V7.0.3 Binary..." + bcolors.ENDC)
+            subprocess.run(["git stash"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
```
Comment (p0mvn) [inline osmosis-installer.py:753]: q: How did it work before without git stash? Why do we need it now?

File: osmosis-installer.py [hunk 2]
```diff
@@ -680,7 +851,7 @@ def initEnvironment():
                 subprocess.run(["clear"], shell=True)
                 initEnvironment()
         else:
-            print(bcolors.OKGREEN +"You have enough RAM to meet the 32GB minimum requirement, moving on to system setup..."+ bcolors.ENDC)
+            print(bcolors.OKGREEN +"You have enough RAM to meet the 64GB minimum requirement, moving on to system setup..."+ bcolors.ENDC)
```
Comment (p0mvn) [inline osmosis-installer.py:854]: q: why have we decided to move on to 64GB recommended?

File: osmosis-installer.py (hunk 1 above)
Comment (czarcas7ic) [reply osmosis-installer.py:753]: Most of the time people ran this script on a fresh VM, and even if they didn't they usually didn't make local changes to the osmosis folder. I only realized after having to directly modify the go.mod to accept rocksdb that you cant automate git checkout if you don't stash first

File: osmosis-installer.py [hunk 3]
```diff
@@ -161,6 +189,142 @@ def cosmovisorInit ():
...
+    print(bcolors.OKGREEN + "Preparing v4 Upgrade..." + bcolors.ENDC)
+    subprocess.run(["git checkout v4.2.0"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True, env=my_env)
+    subprocess.run(["make build"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True, env=my_env)
+    subprocess.run(["cp build/osmosisd "+osmo_home+"/cosmovisor/upgrades/v4/bin"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True, env=my_env)
+    print(bcolors.OKGREEN + "Preparing v5/v6 Upgrade..." + bcolors.ENDC)
+    subprocess.run(["git checkout v6.4.1"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True, env=my_env)
+    subprocess.run(["make build"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True, env=my_env)
+    subprocess.run(["cp build/osmosisd "+osmo_home+"/cosmovisor/upgrades/v5/bin"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True, env=my_env)
+    print(bcolors.OKGREEN + "Preparing v7 Upgrade..." + bcolors.ENDC)
+    subprocess.run(["git checkout v7.0.3"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True, env=my_env)
```
Comment (p0mvn) [inline osmosis-installer.py:232]: nit: 7.0.4 but we can do it in a separate PR

File: osmosis-installer.py (hunk 2 above)
Comment (czarcas7ic) [reply osmosis-installer.py:854]: My 32GB VM that I first tested the replay on got OOMd on one of the upgrades (I think v6?) I really dont think 64 is needed, but I am unclear how far past 32GB it got that made it OOM. What is a solid number that you think we should make this?

File: osmosis-installer.py (hunk 3 above)
Comment (czarcas7ic) [reply osmosis-installer.py:232]: Good catch, thanks

File: osmosis-installer.py [hunk 4]
```diff
@@ -161,6 +189,142 @@ def cosmovisorInit ():
...
+    subprocess.run(["make build"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True, env=my_env)
+    subprocess.run(["cp build/osmosisd "+osmo_home+"/cosmovisor/upgrades/v7/bin"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True, env=my_env)
+    subprocess.run(["git checkout v3.1.0"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True, env=my_env)
```
Comment (p0mvn) [inline osmosis-installer.py:235]: nit: for readability, would be great to do the upgrades in sequence:
1. v3.1.0
2. v4.2.0
3. v7.0.3
4. <future upgrades>

Comment (p0mvn) [inline osmosis-installer.py:235]: Missing: `print(bcolors.OKGREEN + "Preparing v3.1.0 Upgrade..." + bcolors.ENDC)`

Comment (p0mvn) [reply osmosis-installer.py:235]: Nvm, I see why we do it this way

Comment (p0mvn) [reply osmosis-installer.py:235]: Nvm, I see why we do it this way

File: osmosis-installer.py (hunk 2 above)
Comment (p0mvn) [reply osmosis-installer.py:854]: I've never gotten OOM'ed during normal usage of 32GB nodes. However, I've never tested the upgrades.

Do you think it makes sense to require 64GB only if the user selects the option to replay from genesis?

File: osmosis-installer.py [hunk 5]
```diff
@@ -161,6 +189,142 @@ def cosmovisorInit ():
...
+def replayFromGenesisRocksDb ():
+    print(bcolors.OKGREEN + "Changing db_backend to rocksdb..." + bcolors.ENDC)
+    subprocess.run(["sed -i -E 's/db_backend = \"goleveldb\"/db_backend = \"rocksdb\"/g' "+osmo_home+"/config/config.toml"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
+    print(bcolors.OKGREEN + "Installing rocksdb..." + bcolors.ENDC)
+    print(bcolors.OKGREEN + "This process may take 15 minutes or more" + bcolors.ENDC)
+    os.chdir(os.path.expanduser(HOME))
+    subprocess.run(["sudo apt-get install -y libgflags-dev libsnappy-dev zlib1g-dev libbz2-dev liblz4-dev libzstd-dev"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
+    subprocess.run(["git clone https://github.com/facebook/rocksdb.git"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
+    os.chdir(os.path.expanduser(HOME+"/rocksdb"))
+    # originally used v6.27.3
```
Comment (p0mvn) [inline osmosis-installer.py:255]: Does it make sense to keep the comment here if we tested that v6.29.3 works well?

File: osmosis-installer.py [hunk 6]
```diff
@@ -161,6 +189,142 @@ def cosmovisorInit ():
...
+    subprocess.run(["git checkout v6.29.3"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
+    subprocess.run(["export CXXFLAGS='-Wno-error=deprecated-copy -Wno-error=pessimizing-move -Wno-error=class-memaccess'"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
+    subprocess.run(["sudo make shared_lib"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
+    subprocess.run(["sudo make install-shared INSTALL_PATH=/usr"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
+    subprocess.run(["sudo echo 'export LD_LIBRARY_PATH=/usr/local/lib' >> $HOME/.bashrc"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, shell=True)
+    # subprocess.run(["source $HOME/.bashrc"], shell=True)
```
Comment (p0mvn) [inline osmosis-installer.py:261]: Also, not sure if there is a reason for keeping the dead code?

**Review by p0mvn — APPROVED** [review]
LGTM. Just a few minor comments/questions.

Thanks for including me in the reviews. It's been super helpful to learn how things work through them

File: osmosis-installer.py (hunk 2 above)
Comment (czarcas7ic) [reply osmosis-installer.py:854]: Yes, this is a good middle ground, will do this!

File: osmosis-installer.py (hunk 5 above)
Comment (czarcas7ic) [reply osmosis-installer.py:255]: Forgot to remove, thanks

File: osmosis-installer.py (hunk 6 above)
Comment (czarcas7ic) [reply osmosis-installer.py:261]: Nope, will remove now!

**Comment by czarcas7ic** [comment]
Thanks so much for reviewing, will make the recommended changes now and then will merge!
